specs/gloas/beacon-chain.md

```diff
@@ -1048,6 +1146,209 @@ def process_operations(state: BeaconState, body: BeaconBlockBody) -> None:
+    if withdrawal_request.validator_pubkey in builder_pubkeys:
+        index = builder_pubkeys.index(withdrawal_request.validator_pubkey)
+        builder_index = BuilderIndex(index)
+        return process_withdrawal_request_for_builder(state, withdrawal_request, builder_index)
+
+    validator_pubkeys = [v.pubkey for v in state.validators]
+    if withdrawal_request.validator_pubkey in validator_pubkeys:
+        index = validator_pubkeys.index(withdrawal_request.validator_pubkey)
+        validator_index = ValidatorIndex(index)
+        return process_withdrawal_request_for_validator(state, withdrawal_request, validator_index)
+```
+
+##### Deposit requests
+
+###### New `get_index_for_new_builder`
+
+```python
+def get_index_for_new_builder(state: BeaconState) -> BuilderIndex:
+    for index, builder in enumerate(state.builders):
+        if builder.withdrawable_epoch <= get_current_epoch(state) and builder.balance == 0:
+            return BuilderIndex(index)
+    return BuilderIndex(len(state.builders))
+```
+
+###### New `get_builder_from_deposit`
+
+```python
+def get_builder_from_deposit(
+    state: BeaconState, pubkey: BLSPubkey, withdrawal_credentials: Bytes32, amount: uint64
+) -> Builder:
+    return Builder(
+        pubkey=pubkey,
+        execution_address=ExecutionAddress(withdrawal_credentials[12:]),
+        balance=amount,
+        deposit_epoch=get_current_epoch(state),
+        withdrawable_epoch=FAR_FUTURE_EPOCH,
+    )
+```
+
+###### New `add_builder_to_registry`
+
+```python
+def add_builder_to_registry(
+    state: BeaconState, pubkey: BLSPubkey, withdrawal_credentials: Bytes32, amount: uint64
+) -> None:
+    index = get_index_for_new_builder(state)
+    builder = get_builder_from_deposit(state, pubkey, withdrawal_credentials, amount)
+    set_or_append_list(state.builders, index, builder)
+```
+
+###### New `apply_deposit_for_builder`
+
+```python
+def apply_deposit_for_builder(
+    state: BeaconState,
+    pubkey: BLSPubkey,
+    withdrawal_credentials: Bytes32,
+    amount: uint64,
+    signature: BLSSignature,
+) -> None:
+    builder_pubkeys = [b.pubkey for b in state.builders]
+    if pubkey not in builder_pubkeys:
+        # Verify the deposit signature (proof of possession) which is not checked by the deposit contract
+        if is_valid_deposit_signature(pubkey, withdrawal_credentials, amount, signature):
+            add_builder_to_registry(state, pubkey, withdrawal_credentials, amount)
+    else:
+        # Increase balance by deposit amount
+        builder_index = builder_pubkeys.index(pubkey)
+        state.builders[builder_index].balance += amount
+```
+
+###### Modified `process_deposit_request`
+
+```python
+def process_deposit_request(state: BeaconState, deposit_request: DepositRequest) -> None:
+    # Set deposit request start index
+    if state.deposit_requests_start_index == UNSET_DEPOSIT_REQUESTS_START_INDEX:
+        state.deposit_requests_start_index = deposit_request.index
```

Comment:
can this be removed in Gloas?

Reply:
Yes, let’s remove it

Reply:
Removed here: https://github.com/ethereum/consensus-specs/pull/4788/changes/cc04025ae6459f1fafd5dfdee283c026cb15c92e